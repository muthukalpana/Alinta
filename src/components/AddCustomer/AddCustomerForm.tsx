import { Field, Formik, FormikHelpers } from "formik";
import * as React from "react";
import { ICustomer, Customer } from "../../types/types";
import {
  StyledForm,
  StyledInput,
  StyledLabel,
  StyledAddButton,
  ErrorLabel,
} from "./StyledAddCustomerForm";

type Props = {
  customers: readonly ICustomer[];
  saveCustomer: (customer: ICustomer | any) => void;
};

export const AddCustomerForm: React.FC<Props> = ({ customers, saveCustomer }) => {
  return (
    <Formik
      initialValues={{
        firstName: "",
        lastName: "",
        phoneNumber: "",
      }}
      onSubmit={(
        values: Customer,
        { setSubmitting }: FormikHelpers<Customer>
      ) => {
        saveCustomer(values);
        setSubmitting(false);
      }}
    validate={values => {
        const errors = {
          firstName: '',
          phoneNumber: '',
        }
        const nameFilter = customers.find(customer => customer.firstName.toLowerCase().trim() === values.firstName.toLowerCase().trim() &&
        customer.lastName.toLowerCase().trim() === values.lastName.toLowerCase().trim())
        const phoneFilter = customers.find(customer => customer.phoneNumber.trim() === values.phoneNumber.trim())
        if (nameFilter) {
          errors.firstName = `Customer Name (${nameFilter.firstName} ${nameFilter.lastName}) already exists!`
        }
        if (values.phoneNumber !== '') {
          if (phoneFilter) {
            errors.phoneNumber = 'Phone number already exists'
          } else {
            const regex = new RegExp(/^(\+?\(61\)|\(\+?61\)|\+?61|\(0[1-9]\)|0[1-9])?( ?-?[0-9]){7,9}$/)
            const isValid = regex.test(values.phoneNumber)
            errors.phoneNumber = isValid ? '' : 'Please enter valid phone number';
          }
        }
        return errors.phoneNumber === '' && errors.firstName === '' ? {} : errors;
      }}
    >
      {
        (props) => (
          <StyledForm onSubmit={props.handleSubmit}>
            <StyledLabel htmlFor="firstName">First Name</StyledLabel>
            <Field
              as={StyledInput}
              id="firstName"
              name="firstName"
              placeholder="John"
            />

            <StyledLabel htmlFor="lastName">Last Name</StyledLabel>
            <Field
              as={StyledInput}
              id="lastName"
              name="lastName"
              placeholder="Doe"
            />
            { props.errors.firstName !== '' && <ErrorLabel>{ props.errors.firstName }</ErrorLabel>}
            <StyledLabel htmlFor="phoneNumber">Phone Number</StyledLabel>
            <Field
              as={StyledInput}
              id="phoneNumber"
              name="phoneNumber"
              placeholder="Enter number"
              type="tel"
            />
            { props.errors.phoneNumber !== '' && <ErrorLabel>{ props.errors.phoneNumber }</ErrorLabel>}
            <StyledAddButton
              type="submit"
              disabled={props.values.firstName === '' || props.values.lastName === '' || 
                props.values.phoneNumber === '' || (props.errors.phoneNumber !== undefined && props.errors.phoneNumber !== '')
                || (props.errors.firstName !== undefined && props.errors.firstName !== '')}
            >
              Add Customer
            </StyledAddButton>
          </StyledForm>
        )
      }
      </Formik>
  );
};
