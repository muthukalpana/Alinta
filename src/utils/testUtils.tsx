import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { customerReducer } from "../redux/reducers/customerReducers";

// @TODO clean up types here
// Set-up copied from https://redux.js.org/recipes/writing-tests#components
function render(
  // @ts-ignore
  ui,
  {
    // @ts-ignore
    initialState,
    store = createStore(customerReducer, initialState),
    ...renderOptions
  } = {}
) {
  // @ts-ignore
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  // @ts-ignore
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from "@testing-library/react";
// override render method
export { render };
