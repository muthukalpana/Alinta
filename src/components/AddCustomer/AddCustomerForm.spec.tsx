import * as React from "react";
import { render } from "../../utils/testUtils";
import { AddCustomerForm } from "./AddCustomerForm";

const customers = [{
  id: 1,
  firstName: "Test",
  lastName: "Tester",
  phoneNumber: "00000000",
}];

describe("<AddCustomerForm />", () => {
  it("should render a <AddCustomerForm />", () => {
    const wrapper = render(<AddCustomerForm customers={customers} saveCustomer={() => {}} />);
    expect(wrapper.container).toMatchSnapshot();
  });

  //@TODO Add tests for entering data and clicking submit
});
