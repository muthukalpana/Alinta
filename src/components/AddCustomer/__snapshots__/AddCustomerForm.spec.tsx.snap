// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCustomerForm /> should render a <AddCustomerForm /> 1`] = `
<div>
  <form
    action="#"
    class="sc-bdfBQB IjRdz"
  >
    <label
      class="sc-gsTEea ksWhXL"
      for="firstName"
    >
      First Name
    </label>
    <input
      class="sc-dlfnuX eytFzo"
      id="firstName"
      name="firstName"
      placeholder="John"
      value=""
    />
    <label
      class="sc-gsTEea ksWhXL"
      for="lastName"
    >
      Last Name
    </label>
    <input
      class="sc-dlfnuX eytFzo"
      id="lastName"
      name="lastName"
      placeholder="Doe"
      value=""
    />
    <label
      class="sc-eCstlR limRGo"
    />
    <label
      class="sc-gsTEea ksWhXL"
      for="phoneNumber"
    >
      Phone Number
    </label>
    <input
      class="sc-dlfnuX eytFzo"
      id="phoneNumber"
      name="phoneNumber"
      placeholder="Enter number"
      type="tel"
      value=""
    />
    <label
      class="sc-eCstlR limRGo"
    />
    <button
      class="sc-hKgJUU bAjZNY"
      disabled=""
      type="submit"
    >
      Add Customer
    </button>
  </form>
</div>
`;
